### PR TITLE
[REVIEW] Add dictionary support to cudf::unary_operation

### PR DESCRIPTION
**Pull request by davidwendt**
Reference #5963

This PR adds dictionary specialization logic to the `cudf::unary_operation` API.
All math and bitwise operations return new dictionary columns. Logical operations return BOOL8 type columns.
This includes a reworked/simplified `math_ops.cu` and removal of `unary_ops.cuh` which is not needed.
Also, the gtests `unary_ops_test.cpp` math/logical operations were split out into a new `math_ops_test.cpp` to simplify updating tests.

### CODE REVIEW

File: cpp/src/unary/math_ops.cu [hunk 1]
```diff
@@ -229,89 +226,252 @@ struct DeviceNot {
...
+  };
+
+  template <typename T,
+            typename std::enable_if_t<!is_supported<T>() and
+                                      std::is_same<T, dictionary32>::value>* = nullptr>
+  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+  {
+    if (input.size() == 0) return make_empty_column(cudf::data_type{cudf::type_id::BOOL8});
```
Comment (codereport) [inline cpp/src/unary/math_ops.cu:455]:
```suggestion
    if (input.is_empty()) return make_empty_column(cudf::data_type{cudf::type_id::BOOL8});
```

File: cpp/src/unary/math_ops.cu [hunk 2]
```diff
@@ -229,89 +226,252 @@ struct DeviceNot {
...
+  };
+
+  template <typename T,
+            typename std::enable_if_t<!std::is_arithmetic<T>::value and
+                                      std::is_same<T, dictionary32>::value>* = nullptr>
+  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+  {
+    if (input.size() == 0) return empty_like(input);
```
Comment (codereport) [inline cpp/src/unary/math_ops.cu:321]:
```suggestion
    if (input.is_empty()) return empty_like(input);
```

File: cpp/src/unary/math_ops.cu [hunk 3]
```diff
@@ -229,89 +226,252 @@ struct DeviceNot {
...
+  };
+
+  template <typename T,
+            typename std::enable_if_t<!std::is_integral<T>::value and
+                                      std::is_same<T, dictionary32>::value>* = nullptr>
+  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+  {
+    if (input.size() == 0) return empty_like(input);
```
Comment (codereport) [inline cpp/src/unary/math_ops.cu:378]:
```suggestion
    if (input.is_empty()) return empty_like(input);
```

**Review by codereport — CHANGES_REQUESTED** [review]
Just some small first pass changes. Looks good in general though :+1:

File: cpp/src/unary/math_ops.cu [hunk 4]
```diff
@@ -229,89 +226,252 @@ struct DeviceNot {
...
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
 {
-  return cudf::unary::launcher<T, T, F>::launch(input, op, mr, stream);
+  auto dictionary_view = cudf::column_device_view::create(input.parent(), stream);
+  auto keys_view       = cudf::column_device_view::create(input.keys(), stream);
+  auto dictionary_itr  = thrust::make_permutation_iterator(
+    keys_view->begin<T>(),
+    thrust::make_transform_iterator(thrust::make_counting_iterator<size_type>(0),
+                                    access_index_fn{*dictionary_view}));
```
Comment (jrhemstad) [inline cpp/src/unary/math_ops.cu:268]: This seems like a common utility we'd want to reuse elsewhere. Maybe something like a `make_dictionary_iterator(dictionary_column_view const& input)`.

**Review by karthikeyann — CHANGES_REQUESTED** [review]

**Review by codereport — APPROVED** [review]
LGTM :+1: 

There is some refactoring in the unary op tests that I would do - but it is out of scope for this PR (and it is my fault because I wrote the original code) - will set up a follow up PR.

**Comment by codereport** [comment]
rerun tests

**Review by karthikeyann — APPROVED** [review]

File: cpp/include/cudf/dictionary/detail/iterator.cuh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * @brief Create dictionary iterator that produces key elements.
+ *
+ *
+ * @throw cudf::logic_error if `dictionary_column` is not a dictionary column.
+ *
+ * @param dictionary_column The dictionary device view to iterate.
+ * @return Iterator
+ */
+template <typename KeyType>
+auto make_dictionary_iterator(column_device_view const& dictionary_column)
```
Comment (jrhemstad) [inline cpp/include/cudf/dictionary/detail/iterator.cuh:48]: Should this variant should only support columns w/o nulls? I'd imagine we'd have a `pair_iterator` equivalent that can be used for columns with null elements.

Comment (davidwendt) [reply cpp/include/cudf/dictionary/detail/iterator.cuh:48]: I was planning to add the pair-iterator here once I need it. Should be soon but not in this PR.

Comment (jrhemstad) [reply cpp/include/cudf/dictionary/detail/iterator.cuh:48]: Makes sense. I'd think for this PR that this iterator should `CUDF_EXPECT` that the input column doesn't contain nulls. I think similar iterators do the same thing.

Comment (davidwendt) [reply cpp/include/cudf/dictionary/detail/iterator.cuh:48]: Yes, good point. It also seems a little weak on documentation too now that I look a back at it.

Comment (davidwendt) [reply cpp/include/cudf/dictionary/detail/iterator.cuh:48]: The `column_device_view` does not have a `null_count()` nor a `has_nulls()` member so this utility is unable to check. The iterator handles nulls either way so the caller can do the check if necessary.

File: cpp/src/unary/math_ops.cu [hunk 6]
```diff
@@ -229,89 +227,239 @@ struct DeviceNot {
...
+    std::unique_ptr<cudf::column> operator()(cudf::dictionary_column_view const& input,
+                                             rmm::mr::device_memory_resource* mr,
+                                             cudaStream_t stream)
+    {
+      CUDF_FAIL("dictionary keys type not supported for this operation");
+    }
+  };
+
+  template <typename T,
+            typename std::enable_if_t<!std::is_integral<T>::value and
```
Comment (nvdbaranec) [inline cpp/src/unary/math_ops.cu:365]: Does this need to be here?  I'm guessing a dictionary doesn't check out as `is_integral`

**Review by nvdbaranec — APPROVED** [review]
Looks good.  Just had one question about a SFINAE thing.

File: cpp/src/unary/math_ops.cu (hunk 6 above)
Comment (davidwendt) [reply cpp/src/unary/math_ops.cu:365]: Yes, dictionary is not integral.

**Review by harrism — APPROVED** [review]
cmake approval

**Pull request merged**